src/Xamarin.Android.Build.Tasks/Utilities/MonoAndroidHelper.cs

```diff
@@ -363,32 +363,38 @@ public static bool IsForceRetainedAssembly (string assembly)
 
 		public static void SetWriteable (string source)
 		{
-			if (!File.Exists (source))
+			FileAttributes fromAttr;
+			try {
+				fromAttr = File.GetAttributes (source);
+			} catch (FileNotFoundException) {
+				// FNFE should be rare, callers are already checking File.Exists()
 				return;
+			}
 
-			var fileInfo = new FileInfo (source);
-			if (fileInfo.IsReadOnly)
-				fileInfo.IsReadOnly = false;
+			var toAttr = fromAttr & ~FileAttributes.ReadOnly;
+			if (fromAttr != toAttr) {
+				File.SetAttributes (source, toAttr);
+			}
 		}
 
 		public static void SetDirectoryWriteable (string directory)
 		{
-			if (!Directory.Exists (directory))
+			var root = new DirectoryInfo (directory);
+			if (!root.Exists)
 				return;
+			SetWriteable (root);
 
-			var dirInfo = new DirectoryInfo (directory);
-			if ((dirInfo.Attributes | FileAttributes.ReadOnly) == FileAttributes.ReadOnly)
-				dirInfo.Attributes &= ~FileAttributes.ReadOnly;
-
-			foreach (var dir in Directory.EnumerateDirectories (directory, "*", SearchOption.AllDirectories)) {
-				dirInfo = new DirectoryInfo (dir);
-				if ((dirInfo.Attributes | FileAttributes.ReadOnly) == FileAttributes.ReadOnly)
-					dirInfo.Attributes &= ~FileAttributes.ReadOnly;
+			foreach (FileSystemInfo child in root.EnumerateFileSystemInfos ("*", SearchOption.AllDirectories)) {
```

Comment:
If the concern is GC use and reducing the garbage created, wouldn't we want to *avoid* the `DirectoryInfo` creation and use of `root.EnumerateFileSystemInfos()` here?

Reply:
What is the alternative? `DirectoryInfo` is the only way I know how to set directory attributes besides p/invoke. There is not a `Directory.SetAttributes` method.

Reply:
Indeed.  Ugh.